### PR TITLE
Create a fuzz test that randomizes topology changes, cluster actions, and CQL operations

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/impl/INodeProvisionStrategy.java [hunk 1]
```diff
@@ -29,137 +29,80 @@
 @Shared(inner = INTERFACES)
 public interface INodeProvisionStrategy
 {
-    enum Strategy
+    @FunctionalInterface
+    interface Factory
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/INodeProvisionStrategy.java]: this lets me control this interface a lot more in my test... right now the code assumes node1 never goes away, but my test invalidates that assumption... so seed selection and more must live with my test

File: test/unit/accord/utils/Gen.java [hunk 2]
```diff
@@ -87,6 +89,22 @@ default Gen<A> filter(Predicate<A> fn)
         };
     }
 
+    default Gen<A> filter(int maxAttempts, A defaultValue, Predicate<A> fn)
```
Comment (dcapwell) [inline test/unit/accord/utils/Gen.java]: a lot of changes to these classes are not *needed* for this patch, but my thinking is to try to make in-sync with accord so the next rebase isn't a pain

File: test/unit/accord/utils/Gens.java [hunk 3]
```diff
@@ -19,26 +19,39 @@
 package accord.utils;
 
 import java.lang.reflect.Array;
+import java.math.BigDecimal;
```
Comment (dcapwell) [inline test/unit/accord/utils/Gens.java]: a lot of changes to these classes are not *needed* for this patch, but my thinking is to try to make in-sync with accord so the next rebase isn't a pain

File: test/unit/accord/utils/Gens.java [hunk 4]
```diff
@@ -89,6 +145,215 @@ public static <T> Gen<T> pick(Map<T, Integer> values)
...
+                array[i] = minInclusive + i * delta;
+            indexes = IntStream.range(0, array.length).toArray();
+        }
+        else
+        {
+            array = IntStream.range(minInclusive, maxExclusive).toArray();
+            indexes = null;
+        }
+        return rs -> {
+            switch (rs.nextInt(0, 2))
```
Comment (dcapwell) [inline test/unit/accord/utils/Gens.java]: this isn't 100% in-sync with accord due to a missing set of random methods I didn't care to backport... so this is missing 2 other "distributions"

File: test/unit/accord/utils/Gens.java [hunk 5]
```diff
@@ -433,16 +735,7 @@ public Gen<T[]> ofSizeBetween(int minSize, int maxSize) {
...
-                    try
-                    {
-                        list[i] = fn.next(r);
-                    }
-                    catch (IgnoreGenResult e)
-                    {
-                        return Arrays.copyOf(list, i);
-                    }
-                }
+                    list[i] = fn.next(r);
```
Comment (dcapwell) [inline test/unit/accord/utils/Gens.java]: I don't know when/why this was removed and looking at the `List` version this looks like a bug... we don't hit this case often with arrays which is prob why we didn't notice...

I plan to create a patch for cep-15-accord + accord to keep them in-sync with this PR... ill try to mixup in both places soon

File: test/unit/accord/utils/Property.java [hunk 6]
```diff
@@ -18,23 +18,25 @@
...
 import javax.annotation.Nullable;
 
-import org.apache.cassandra.transport.ProtocolException;
-import org.apache.cassandra.utils.concurrent.AsyncPromise;
-
 public class Property
 {
     public static abstract class Common<T extends Common<T>>
     {
-        protected long seed = ThreadLocalRandom.current().nextLong();
```
Comment (dcapwell) [inline test/unit/accord/utils/Property.java]: `ThreadLocalRandom` should be avoided at all costs... the seed isn't random which means restarting the JVM yields the same seed... so you keep running the same tests over and over and over again...

File: test/unit/org/apache/cassandra/utils/ConfigGenBuilder.java [hunk 7]
```diff
@@ -70,7 +70,7 @@ public enum Memtable
     /**
      * When loading the {@link Config} from a yaml its possible that some configs set will conflict with the configs that get generated here, to avoid that set them to a good default
      */
-    public static Config santize(Config config)
+    public static Config sanitize(Config config)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/ConfigGenBuilder.java]: @maedhroz you missed a typo!  SHAME!!!!!

File: src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java [hunk 8]
```diff
@@ -56,7 +56,7 @@ public final Commit.Result commit(Entry.Id entryId, Transformation transform, fi
         {
             ClusterMetadata previous = log.waitForHighestConsecutive();
             if (!previous.fullCMSMembers().contains(FBUtilities.getBroadcastAddressAndPort()))
-                throw new IllegalStateException("Node is not a member of CMS anymore");
+                throw new IllegalStateException("Node " + FBUtilities.getBroadcastAddressAndPort() + " is not a member of CMS anymore in " + previous.epoch + "; members=" + previous.fullCMSMembers());
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java]: this was super flakey and added a ignore filter so the test won't fail if this is seen anymore (assuming its safe... think it is)...

I saw a history where node1 starts a snapshot, node1 looses its CMS membership (nodes 2/3/4 are now the members), then this check blows up...

The extra details just help know *when* in the timeline things happened...

File: test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        private int numInserts = 0;
+        public HarryState(RandomSource rs)
+        {
+            super(rs, HarryTopologyMixupTest::createSchemaSpec, HarryTopologyMixupTest::cqlOperations);
+        }
+
+        @Override
+        protected void onConfigure(IInstanceConfig config)
+        {
+            config.set("metadata_snapshot_frequency", 5);
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java]: I don't remember why I changed this... I have yet to find a history for *why*

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+            removeTypeGen = REMOVE_TYPE_DISTRIBUTION.next(rs);
+
+            long waitForEpoch = HackSerialization.tcmEpoch(cluster.get(1));
+            currentEpoch.set(waitForEpoch);
+            onStartupComplete(waitForEpoch);
+        }
+
+        protected void onStartupComplete(long tcmEpoch)
+        {
+
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: not used in this patch but accord needs to block on accord knowing this epoch

File: src/java/org/apache/cassandra/tcm/ClusterMetadataService.java [hunk 11]
```diff
@@ -844,11 +844,22 @@ private Pair<State, Processor> delegateInternal()
...
+                    if (delegate.left == LOCAL || delegate.left == RESET)
+                        replicator.send(result, null);
+                    return result;
+                }
+                catch (NotCMSException e)
+                {
+                    retryPolicy.maybeSleep();
+                }
+            }
+            return Commit.Result.failed(ExceptionCode.SERVER_ERROR, "Could not commit " + transform.kind() + " at epoch " + lastKnown);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ClusterMetadataService.java]: nit: should we use string format here?

File: src/java/org/apache/cassandra/tcm/ClusterMetadataService.java [hunk 12]
```diff
@@ -844,11 +844,22 @@ private Pair<State, Processor> delegateInternal()
         @Override
         public Commit.Result commit(Entry.Id entryId, Transformation transform, Epoch lastKnown, Retry.Deadline retryPolicy)
         {
-            Pair<State, Processor> delegate = delegateInternal();
-            Commit.Result result = delegate.right.commit(entryId, transform, lastKnown, retryPolicy);
-            if (delegate.left == LOCAL || delegate.left == RESET)
-                replicator.send(result, null);
-            return result;
+            while (!retryPolicy.reachedMax())
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ClusterMetadataService.java]: I _think_ we should not retry here, and instead should retry in an underlying processor. But maybe this is one of the Marcus's patches.

File: test/distributed/org/apache/cassandra/distributed/shared/ClusterUtils.java [hunk 13]
```diff
@@ -611,8 +632,8 @@ public static void waitForCMSToQuiesce(ICluster<IInvokableInstance> cluster, Epo
                 if (cluster.get(j).isShutdown())
                     continue;
                 Epoch version = getClusterMetadataVersion(cluster.get(j));
-                if (!awaitedEpoch.equals(version))
-                    notMatching.add(new ClusterMetadataVersion(j, version, getClusterMetadataVersion(cluster.get(j))));
+                if (!awaitedEpoch.equals(version) && version.getEpoch() < awaitedEpoch.getEpoch())
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/shared/ClusterUtils.java]: Just `version.getEpoch() < awaitedEpoch.getEpoch()` should be enough here, or?

File: test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        return (rs, state) -> {
+            HarryState harryState = (HarryState) state;
+            TopologyHistory history = state.topologyHistory;
+            // if any topology change happened, then always validate all
+            if (harryState.generation != history.generation())
+            {
+                harryState.generation = history.generation();
+                return validateAll;
+            }
+            if (harryState.numInserts > 500
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java]: I do not mind the current state, but am wondering if a slightly different interface might be more convenient? One option would be to use something like "stateful model checker" from QT, where you would specify the possible commands, and supply their probabilities (i.e. 0.8 / 0.2 writes to reads distribution), something like a `RangeTombstoneBurnTest` would do:

```
var insertChance = 0.8;
var readChacne = 0.2;
                            .step((history, rng) -> rng.nextFloat() > insertChance,
                                  (history, rng) -> {
...
                                  })
                            .step((history, rng) -> rng.nextFloat() > readChance,
                                  (history, rng) -> {
..
                                  })
```

This interface also nicely fits with the `State` idea you got here.

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 15]
```diff
@@ -0,0 +1,756 @@
...
+            init(cluster, TARGET_RF);
+            // fix TCM
+            {
+                NodeToolResult result = cluster.get(1).nodetoolResult("cms", "reconfigure", "2");
+                result.asserts().success();
+                logger.info("CMS reconfigure: {}", result.getStdout());
+            }
+            preActions.add(new Runnable()
+            {
+                // in order to remove this action, an annonomus class is needed so "this" works, lambda "this" is the parent class
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: annonomus > anonymous

Comment (ifesdjeen) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: Also, if you like you can use `preActions.remove(State.this);` and lambda. I don't mind an anonymous class though

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 16]
```diff
@@ -0,0 +1,756 @@
...
+        {
+            Node n = nodes.get(i);
+            if (n == null)
+                throw new IllegalArgumentException("Unknown node" + i);
+            return n.tokens;
+        }
+
+        public int[] active()
+        {
+            int[] keys = new int[nodes.size()];
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: this one seems to be returning all nodes not just active, or?

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 17]
```diff
@@ -0,0 +1,756 @@
...
+
+        public int size()
+        {
+            return nodes.size();
+        }
+
+        public Node addNode()
+        {
+            int id = ++uniqueInstances;
+            List<String> instTokens = Gens.lists(MURMUR_TOKEN_GEN
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: should we make this static? looks like we want just one instance of this generator, or?

**Review by ifesdjeen — COMMENTED** [review]
I like the direction where this is headed and am mostly on board with changes. The only thing I wanted to propose is a possible modification in the API. Rather than pre-declaring possible commands that could potentially be executed, we could do something like:

```
                            .command((state, rng) -> {
                                      rng.nextFloat() > threshold && // compute chance
                                      predicate.test(state)          // precondition (for example, do we have enough node to even start decomission 
                                  },
                                  (state, rng) -> {
                                      return multistep(...
                                  })
                            .command((state, rng) -> {
                                      rng.nextFloat() > threshold && // compute chance
                                      predicate.test(state)          // precondition (for example, do we have enough node to even start decomission 
                                  },
                                  (state, rng) -> {
                                      return multistep(...
                                  })
```

WDYT?

File: src/java/org/apache/cassandra/tcm/ClusterMetadataService.java (hunk 12 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tcm/ClusterMetadataService.java]: I believe im reverting this as Marcus is pushing a fix in his own JIRA

File: src/java/org/apache/cassandra/tcm/ClusterMetadataService.java (hunk 11 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tcm/ClusterMetadataService.java]: I believe im reverting this as Marcus is pushing a fix in his own JIRA

File: test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java (hunk 14 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java]: >  like "stateful model checker" from QT

where is this?  your example is harry's and the QT one Jordan worked on never merged nor really got used in the project.

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java]: your comment is a "builder" of `Command`?  In this case I don't feel it makes too much sense as we have another thing to address

```
            if (harryState.generation != history.generation())
            {
                harryState.generation = history.generation();
                return validateAll;
            }
```

right now the logic says 20% of the time we do reads *or* we have done 500 inserts (honestly copy/paste from harry code... no issues dropping ^_^).

TL;DR - this `Command` does

1) if a topology change happened, always do a read
2) 20% of the time do a read
3) 80% of the time do a write

I am not sure how a builder would make this easier to read...

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 15 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: > if you like you can use preActions.remove(State.this);

what do you mean?  if this becomes a lambda then `this` is `State.this` which is not the correct reference to remove... I would need to save to a variable to keep access *or* make it anonymous...

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 16 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: dead code... this *used* to be different but I removed its usage... just didn't cleanup

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 17 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: not possible to be static.

the first line is

`Gens.lists(MURMUR_TOKEN_GEN.filterAsInt(t -> !activeTokens.contains(Integer.toString(t))))`

this references *current* state, which can't be accessed statically.

second

`.unique()` "could" be static but isn't thread safe... so dangerous to do.

**Comment by dcapwell** [comment]
fixed all feedback or asked for more details (also poked you in slack)

you have 2 similar / related feedbacks, so trying to address here (GH isn't letting me reply...)

https://github.com/apache/cassandra/pull/3486#pullrequestreview-2263041945 - I tested out a `CommandsBuilder` and it made the test `RouteIndextTest` in cep-15-accord easier to read, but I don't see how it helps in this case due to 2 variables that I am not sure how to handle (without making the builder specific to this class)

```
final List<Runnable> preActions = new CopyOnWriteArrayList<>();
final List<Runnable> onError = new CopyOnWriteArrayList<>();
```

The `State` has a `preActions` and `onError` actions... these kinda distort the model and are plugged into the `Commands` functions... if I move to a `CommandsBuilder` model I think `onError` doesn't really change much, but `preActions` gets funky.  Personally I feel the `preActions` is a hack... it would be great to return `Command` so the history does reflect those steps... but that would kinda require each `Command` returned to be a `MultiCommand`... and what about actions that don't mutate anything and just update the `State`'s bookkeeping?  those don't change the history and would just clutter up the error message...


https://github.com/apache/cassandra/pull/3486#discussion_r1732701880

I don't see how a builder works in this specific case.  We are creating a single `Command` and this command has 3 behaviors: if topology changed do read, 20% of the time do reads, 80% of the time do writes.... the builder would need to handle the `if topology changed` bit... which seems funky... and doing a builder for the 20%/80% case seems more verbose than the current logic

**Comment by dcapwell** [comment]
ok pushed to using the new `CommandsBuilder`... the test is now

```
statefulBuilder.check(commands(this::stateGen)
                              .preCommands(state -> state.preActions.forEach(Runnable::run))
                              .add(2, (rs, state) -> {
                                  EnumSet<TopologyChange> possibleTopologyChanges = possibleTopologyChanges(state);
                                  if (possibleTopologyChanges.isEmpty()) return ignoreCommand();
                                  return topologyCommand(state, possibleTopologyChanges).next(rs);
                              })
                              .add(1, (rs, state) -> repairCommand(state, rs.pickInt(state.topologyHistory.up())))
                              .add(7, (rs, state) -> state.statementGen.apply(rs, state))
                              .destroyState((state, cause) -> {
                                  TopologyMixupTestBase.this.destroyState(state, cause);
                                  state.close();
                              })
                              .build());
```

still not a fan of `preCommands` but also not a fan of what I am doing now... at least this interface doesn't make things worse...

one thing that is nice with this builder is I flip `Gen<Command>, Weight`, I personally prefer weight first then the command, always found the gen first harder to read (its an impl detail, we store the Gen in the map as the key)

**Comment by dcapwell** [comment]
also removed `onError` as that is an accord thing... I added that for better debugging of timeout... it doesn't need to be in the base interface and can be done in `AccordTopologyMixupTest.destroyState`, so moving it there

File: src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java (hunk 8 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java]: I am going to drop as Marcus is making similar change in his patch....

Comment (dcapwell) [reply src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java]: https://issues.apache.org/jira/browse/CASSANDRA-19872

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 18]
```diff
@@ -0,0 +1,697 @@
...
+            try
+            {
+
+                this.yamlConfigOverrides = CONF_GEN.next(rs);
+                cluster = Cluster.build(topologyHistory.minNodes)
+                                 .withTokenSupplier(topologyHistory)
+                                 .withConfig(c -> {
+                                     c.with(Feature.values())
+                                      .set("write_request_timeout", "10s");
+                                     //TODO (mainatiance): where to put this?  Anything touching ConfigGenBuilder with jvm-dtest needs this...
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: Nit: mainatiance (and maintnance) below :)

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 19]
```diff
@@ -0,0 +1,697 @@
...
+                                     ((InstanceConfig) c).remove("commitlog_sync_period_in_ms");
+                                     for (Map.Entry<String, Object> e : yamlConfigOverrides.entrySet())
+                                         c.set(e.getKey(), e.getValue());
+                                     onConfigure(c);
+                                 })
+                                 //TODO (maintance): should TopologyHistory also be a INodeProvisionStrategy.Factory so address information is stored in the Node?
+                                 //TODO (maintance): AbstractCluster's Map<Integer, NetworkTopology.DcAndRack> nodeIdTopology makes playing with dc/rack annoying, if this becomes an interface then TopologyHistory could own
+                                 .withNodeProvisionStrategy((subnet, portMap) -> new INodeProvisionStrategy.AbstractNodeProvisionStrategy(portMap)
+                                 {
+                                     {
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: nit: probably extra brackets are not necessary

File: test/unit/accord/utils/async/TimeoutUtils.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+public class TimeoutUtils
+{
+    public interface FailingRunnable
+    {
+        void run() throws Throwable;
+    }
+
+    public static void runBlocking(Duration timeout, String threadName, FailingRunnable fn) throws ExecutionException, InterruptedException, TimeoutException
+    {
+        // MAINTAINCE: Once the accord branch merges to trunk this can be dropped and will be AsyncChain again, but since this is forked into C* (that doesn't have AsyncChain) need to use Futures
```
Comment (ifesdjeen) [inline test/unit/accord/utils/async/TimeoutUtils.java]: nit: MAINTAINCE

**Review by ifesdjeen — APPROVED** [review]

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 19 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: in this case not true.  This is the constructor
